contracts/MarketOracle.sol

```diff
@@ -3,149 +3,244 @@ pragma solidity 0.4.24;
 import "openzeppelin-solidity/contracts/math/SafeMath.sol";
 import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
 
-import "./MarketSource.sol";
+import "./lib/Select.sol";
+
+
+interface IOracle {
+    function getData() external returns (uint256, bool);
+}
 
 
 /**
- * @title Market Oracle
+ * @title Median Oracle
  *
- * @dev Provides the exchange rate and volume data onchain using data from a whitelisted
- *      set of market source contracts.
- *      Exchange rate is the TOKEN:TARGET rate.
- *      Volume is a 24 hour trading volume in Token volume.
+ * @notice Provides a value onchain that's aggregated from a whitelisted set of
+ *         providers.
  */
-contract MarketOracle is Ownable {
+contract MedianOracle is Ownable, IOracle {
     using SafeMath for uint256;
 
-    // Whitelist of sources
-    MarketSource[] public whitelist;
+    struct Report {
+        uint256 timestamp;
+        uint256 payload;
+    }
+
+    // Addresses of providers authorized to push reports.
+    address[] public providers;
+
+    // Reports indexed by provider address. Report[0].timestamp > 0
+    // indicates provider existence.
+    mapping (address => Report[2]) public providerReports;
+
+    event ProviderAdded(address provider);
+    event ProviderRemoved(address provider);
+    event ReportTimestampOutOfRange(address provider);
+
+    // The number of seconds after which the report is deemed expired.
+    uint256 public reportExpirationTimeSec;
 
-    event LogSourceAdded(MarketSource source);
-    event LogSourceRemoved(MarketSource source);
-    event LogSourceExpired(MarketSource source);
+    // The number of seconds since reporting that has to pass before a report
+    // is usable.
+    uint256 public reportDelaySec;
+
+    // The minimum number of providers with valid reports to consider the
+    // aggregate report valid.
+    uint256 public minimumProviders = 1;
+
+    // Timestamp of 1 is used to mark uninitialized and invalidated data.
+    // This is needed so that timestamp of 1 is always considered expired.
+    uint256 private constant MAX_REPORT_EXPIRATION_TIME = 10 years;
```

Comment:
Remember `years` throwing a compilation warning. Maybe move to `days`?

Reply:
Done.